cellmlmanip/singularity_fixes.py

```diff
@@ -0,0 +1,219 @@
+                find_U = d.match(exp_function(U_wildcard) * -Z + 1.0)  # look for exp(U) * -Z + 1.0
+                if not check_bottom_match(find_U):
+                    find_U = d.match(exp_function(U_wildcard) * Z - 1.0)  # look for exp(U) * Z - 1.0
+                if check_bottom_match(find_U):
+                    # We found a match, since exp(U) * Z == exp(U + log(Z)) we can bring Z into the u expression
+                    u = (find_U[U_wildcard] + log(find_U[Z]))
+                    try:
+                        # Find the singularity point by solving u for V==0, excluding irratinal results
+                        sp = tuple(filter(lambda s: not s.has(I), solveset(u, V)))
+                        if sp:
+                            # we found a singularity point, now find vs,ve (or vmin, vmax)
+                            # by solving for U_offset either side of U
+                            find_v_low = solveset(u + U_offset, V)
+                            find_v_up = solveset(u - U_offset, V)
+                            find_v_low = tuple(find_v_low)
+                            find_v_up = tuple(find_v_up)
+                            assert find_v_low and find_v_up and len(find_v_low) == len(find_v_up) == len(sp) == 1, \
+                                'Expecting exactly 1 singularity point '
+                            (vs, ve, sp) = (find_v_low[0], find_v_up[0], sp[0])
+                    except TypeError:
+                        pass  # Result could be 'ConditionSet' which is not iterable and not Real
+
+                if vs is not None:  # check top
+                    for n in num:  # Check arguments in numerator (or denominator)
+                        match = n.match(Z * V - Z * SP_wildcard)  # search for a multiple of V - sp
+                        found_on_top = check_top_match(match, sp)
+                        if not found_on_top:
+                            # search for a exp(multiple of V - sp)
+                            match = n.match(exp_function(Z * V - Z * SP_wildcard))
+                            found_on_top = check_top_match(match, sp)
+                        if found_on_top:  # We've found a match stop looking in the other numerator arguments
+                            break
+                    if found_on_top:  # found singularity, no need to try further
+                        break
+                    else:
+                        (vs, ve, sp) = None, None, None
+        if vs is not None and found_on_top:  # found singularity, no need to try further
+            break
+        else:
+            (vs, ve, sp) = None, None, None
+
+    # Put dummies back in and return the singularity point and range bundries
+    return (float_dummies(vs), float_dummies(ve), float_dummies(sp))
+
+
+def _new_expr_parts(expr, V, U_offset, exp_function):
+    """Removes suitable singularities and replaces it with a piecewise, returning (vs, ve, sp, has_singularity)"""
```

Comment:
If it removes singularities why is it called "new_exp_parts" ??

Reply:
I renamed it but it's tricky. it's a recurive algorithm that eventually removes singularities, but also passes round intermediate reults